File: backend/src/journal/tailer.ts
import { open } from 'node:fs/promises';
import path from 'node:path';
import { parseJournalLine } from './parse.js';
import type { JournalEvent } from '@ed/shared';

export interface TailBatch {
  filename: string;
  events: { lineNo: number; raw: string; event: JournalEvent }[];
  /** byte offset after consuming this batch (only complete lines) */
  newOffset: number;
}

interface FileCursor {
  offset: number; // bytes consumed, always at a line boundary
  lineNo: number; // complete lines consumed so far
}

const NEWLINE = 0x0a;

/**
 * Offset-tracked tailer for journal files. Each poll reads bytes past the
 * cursor and consumes only complete (newline-terminated) lines; a trailing
 * partial line stays on disk and is re-read next poll. Offsets therefore
 * always sit on line boundaries, so a consumer persisting `newOffset` never
 * loses or duplicates events across restarts.
 */
export class JournalTailer {
  private cursors = new Map<string, FileCursor>();

  /** Seed a cursor (e.g. from persisted DB offsets) before the first poll. */
  seed(filename: string, offset: number, lineNo: number): void {
    this.cursors.set(filename, { offset, lineNo });
  }

  has(filename: string): boolean {
    return this.cursors.has(filename);
  }

  /** Read any new complete lines from the file. Returns null if nothing new. */
  async poll(filePath: string): Promise<TailBatch | null> {
    const filename = path.basename(filePath);
    let cursor = this.cursors.get(filename);
    if (!cursor) {
      cursor = { offset: 0, lineNo: 0 };
      this.cursors.set(filename, cursor);
    }

    let fh;
    try {
      fh = await open(filePath, 'r');
    } catch {
      return null; // file vanished between watch event and read
    }
    try {
      const stat = await fh.stat();
      if (stat.size < cursor.offset) {
        // Journal files are append-only; shrinkage means something odd. Re-read.
        cursor.offset = 0;
        cursor.lineNo = 0;
      }
      if (stat.size === cursor.offset) return null;

      const toRead = stat.size - cursor.offset;
      const buf = Buffer.alloc(toRead);
      const { bytesRead } = await fh.read(buf, 0, toRead, cursor.offset);
      const data = buf.subarray(0, bytesRead);

      const events: TailBatch['events'] = [];
      let start = 0;
      for (let i = 0; i < data.length; i++) {
        if (data[i] !== NEWLINE) continue;
        const line = data.subarray(start, i).toString('utf8');
        cursor.lineNo++;
        const event = parseJournalLine(line);
        if (event) events.push({ lineNo: cursor.lineNo, raw: line.trim(), event });
        start = i + 1;
      }
      cursor.offset += start; // only complete lines; partial tail re-read next poll

      if (events.length === 0) return null;
      return { filename, events, newOffset: cursor.offset };
    } finally {
      await fh.close();
    }
  }
}

File: client/tailer.go
package main

import (
	"bytes"
	"encoding/json"
	"io"
	"os"
	"strings"
)

// TailEvent is one journal line to forward, with its physical line number.
type TailEvent struct {
	LineNo int    `json:"lineNo"`
	Raw    string `json:"raw"`
}

// TailResult is the outcome of reading new bytes from a journal file.
type TailResult struct {
	Events    []TailEvent
	NewOffset int64
	NewLineNo int
	// Consumed is how many bytes of complete lines were read (>= 0). Even when
	// no valid events were produced, the offset should advance by this much.
	Consumed int64
}

// validEvent reports whether a journal line is a well-formed event object
// (matches the server's parser: needs string "event" and "timestamp").
func validEvent(line []byte) bool {
	var e struct {
		Event     string `json:"event"`
		Timestamp string `json:"timestamp"`
	}
	if json.Unmarshal(line, &e) != nil {
		return false
	}
	return e.Event != "" && e.Timestamp != ""
}

// tail reads complete (newline-terminated) lines past cur from the file. A
// trailing partial line is left for the next poll, so offsets always sit on
// line boundaries.
func tail(path string, cur Cursor) (TailResult, error) {
	f, err := os.Open(path)
	if err != nil {
		return TailResult{}, err
	}
	defer f.Close()

	st, err := f.Stat()
	if err != nil {
		return TailResult{}, err
	}
	// Journal files are append-only; a smaller size means it was rotated/reset.
	if st.Size() < cur.Offset {
		cur = Cursor{}
	}
	if st.Size() == cur.Offset {
		return TailResult{NewOffset: cur.Offset, NewLineNo: cur.LineNo}, nil
	}

	if _, err := f.Seek(cur.Offset, io.SeekStart); err != nil {
		return TailResult{}, err
	}
	data, err := io.ReadAll(f)
	if err != nil {
		return TailResult{}, err
	}

	res := TailResult{NewLineNo: cur.LineNo}
	lineNo := cur.LineNo
	var consumed int64
	for {
		nl := bytes.IndexByte(data[consumed:], '\n')
		if nl < 0 {
			break // trailing partial line — re-read next poll
		}
		line := data[consumed : consumed+int64(nl)]
		consumed += int64(nl) + 1
		lineNo++
		if validEvent(line) {
			res.Events = append(res.Events, TailEvent{LineNo: lineNo, Raw: strings.TrimSpace(string(line))})
		}
	}
	res.Consumed = consumed
	res.NewOffset = cur.Offset + consumed
	res.NewLineNo = lineNo
	return res, nil
}
